feat: add SP-specific contributing guide (#521)
canonical/sphinx-docs-starter-pack: CONTRIBUTING.md

# Contributing to the starter pack

The starter pack provides a shared foundation for Sphinx-based documentation across multiple Canonical projects. The work done here benefits every team that uses it. While most of the work is done by the Documentation Practice team, this in no way limits the external contributions.

Common contributions include:

- Bug fixes: Build errors, broken links, configuration issues

> - Documentation: Guides for starter pack features, syntax references, troubleshooting, ad-hoc how-to guides

Reviewer: ```suggestion
- Documentation: How-to guides on Starter Pack features and usage, syntax references, troubleshooting information
```
all minor tweaks, totally optional

### Canonical Contributor License Agreement

Please read and sign our [Contributor License Agreement (CLA)](https://ubuntu.com/legal/contributors) before submitting any changes. The agreement grants Canonical permission to use your contributions. The author of a change remains the copyright owner of their code (no copyright assignment occurs).

Before committing anything, review the terms of the agreement. If you agree and sign it, your work can be incorporated into the repository.

#### CLA check in CI

When you open a pull request (PR) against the `main` branch, an automated check verifies that you have signed the CLA. This check uses the [canonical/has-signed-canonical-cla](https://github.com/canonical/has-signed-canonical-cla) GitHub Action.

If you haven't signed the CLA:
1. The check will fail with a message indicating the CLA requirement
2. Visit <https://ubuntu.com/legal/contributors> to review and sign the agreement
3. Once signed, ask a maintainer to re-run the check, or push a new commit to trigger re-evaluation

> The CLA check only runs on PRs to `main`. Internal team members working on other branches should ensure they have signed the CLA before their changes are merged to `main`.

Reviewer: I think this should be moved to the point when it's of concern to the user (probably during review).

At this point in the guide, the user has yet to open an issue, setup their environment, or open a PR. Bringing up how this is eventually going to impact CI is a big diversion.

Maybe we could instead say something like:

> This is required for all contributors, and will be confirmed during the review process.

That wording could be improved, but something to that effect.

### Open source license

The starter pack is licensed under [GPL-3.0](LICENSE). Documentation for this project is licensed under CC-BY-SA 3.0.

## Report an issue or open a request

> If you find a bug or feature gap in the starter pack, look for it in the [project's GitHub issues](https://github.com/canonical/sphinx-docs-starter-pack/issues) first. If you have fresh input, add your voice to share new findings.

Reviewer: ```suggestion
If you find a bug or feature gap in the starter pack, look for it in the [project's GitHub issues](https://github.com/canonical/sphinx-docs-starter-pack/issues) first. Add your voice to the thread if you have fresh input.
```

"new findings" and "fresh input" feel redundant.

When in doubt, open an issue first to discuss whether the change aligns with the project's goals.

## Development setup

Create a [personal fork](https://github.com/canonical/sphinx-docs-starter-pack/fork) of the repository, then clone it and add the upstream remote:

With [SSH](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/adding-a-new-ssh-key-to-your-github-account):

```bash
git clone git@github.com:<username>/sphinx-docs-starter-pack
cd sphinx-docs-starter-pack
git remote add upstream git@github.com:canonical/sphinx-docs-starter-pack
git fetch upstream
```

With [HTTPS](https://docs.github.com/en/get-started/git-basics/about-remote-repositories#cloning-with-https-urls):

```bash
git clone https://github.com/<username>/sphinx-docs-starter-pack
cd sphinx-docs-starter-pack
git remote add upstream https://github.com/canonical/sphinx-docs-starter-pack
git fetch upstream
```

Install dependencies and verify the build:

```bash
cd docs
make install
make html
```

## Contribute a change

### Research the topic

All significant work should be tied to an existing issue. Before starting, comment on the issue to have it assigned to you.

#### Minor changes

> Check [GitHub issues](https://github.com/canonical/sphinx-docs-starter-pack/issues) for existing reports. If none exists, [open one](https://github.com/canonical/sphinx-docs-starter-pack/issues/new/choose) and indicate you'd like to work on it.

Reviewer: A typo and a very subjective suggestion
```suggestion
Check [GitHub issues](https://github.com/canonical/sphinx-docs-starter-pack/issues) for existing reports. If none exist, [open one](https://github.com/canonical/sphinx-docs-starter-pack/issues/new/choose) and state your interest in working on it.
```

#### Major changes

Describe your proposal in the issue thread, including the plan, tests, and documentation. For new documentation pages, propose a [Diataxis](https://diataxis.fr) category.

### Create a development branch

Sync and create a new branch:

```bash
git fetch upstream
git checkout -b <new-branch-name> upstream/dev

> ```

Reviewer: ````suggestion
```bash
git fetch upstream
git checkout -b <new-branch-name>
```
````
The new branching, versioning, and release policies will save us from needing the `dev` branch. The 1.5.0 release should be the last to use this approach.

1. Amend the most recent commit by signing it without changing the commit message, and push again:

   ```bash
   git commit --amend --no-edit -n -S
   git push
   ```

2. If you still encounter the same error, confirm that your GitHub account has been set up properly to sign commits as described in the [GitHub Docs - About commit signature verification](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification).

   > Tip
   >
   > If you use SSH keys to sign your commits, make sure to add a "Signing Key" type in your GitHub account. See [GitHub Docs - Adding a new SSH key to your account](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/adding-a-new-ssh-key-to-your-github-account) for more information.

### Test the change

Build and run the checks locally before submitting:

```bash
cd docs
make html
```

```bash
make spelling      # Check spelling
make linkcheck     # Validate links
make woke          # Check inclusive language
make lint-md       # Check Markdown style
make vale          # Check style guide compliance (optional)
```

> Preview locally with live reload at `http://127.0.0.1:8000`:

Reviewer: ```suggestion
To instead preview locally with live reload at `http://127.0.0.1:8000`, run:
```

### Ubuntu Code of Conduct

When contributing, you must abide by the [Ubuntu Code of Conduct](https://ubuntu.com/community/ethos/code-of-conduct). Projects governed by Canonical expect good conduct and excellence from every member.

### Canonical Contributor License Agreement

Please read and sign our [Contributor License Agreement (CLA)](https://ubuntu.com/legal/contributors) before submitting any changes. The agreement grants Canonical permission to use your contributions. The author of a change remains the copyright owner of their code (no copyright assignment occurs).

Before committing anything, review the terms of the agreement. If you agree and sign it, your work can be incorporated into the repository.

#### CLA check in CI

When you open a pull request (PR) against the `main` branch, an automated check verifies that you have signed the CLA. This check uses the [canonical/has-signed-canonical-cla](https://github.com/canonical/has-signed-canonical-cla) GitHub Action.

If you haven't signed the CLA:
1. The check will fail with a message indicating the CLA requirement
2. Visit <https://ubuntu.com/legal/contributors> to review and sign the agreement

> 3. Once signed, re-run the failed check or push a new commit to trigger re-evaluation

Reviewer 2: Will contributors who aren't listed as maintainers be able to run checks?

Author: Fair point, I don't know; need to generalize a bit

Run the same basic checks locally that GitHub runs on PRs; see [Test the change](#test-the-change).

#### Changelog guidance

Doc-only changes generally do not require changelog entries. However, reviewers may request one for notable additions such as significant new how-to guides or reference documentation.

For non-documentation changes, ensure that feature changes and fixes are documented in the relevant release notes.

### Push the branch and open a PR

```bash
git push -u origin <branch-name>
```

Next, open a PR on GitHub. Format its title as a conventional commit (GitHub may do this automatically for single-commit branches).

### Describing PRs

Your PR should include the following details:

- Title: Short, descriptive summary
- Description: Problem solved, features added, or bugs fixed
- Relevant issues: [Link related issues and PRs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/autolinked-references-and-urls)
- Testing: How reviewers can verify the change or test the fix
- Reversibility: For costly-to-reverse decisions, explain reasoning and reversal steps

> Make sure to peek at the preview for documentation changes (find the `Read the Docs build` check and click `...` - `View details`).

Author: Good practice, if somewhat circular. OK.

## CI/CD pipeline

The repository configures multiple automated checks. Some are conditional based on target branch or changed files.

> Workflows marked "callable" below support `workflow_call`; those marked "manual dispatch" support `workflow_dispatch` via the GitHub UI.

Reviewer 2: I suggest removing this sentence and the indications of which workflows are callable or support manual dispatch. It doesn't feel crucial to summarise here, and also feels like a pain to maintain.

Author: OK.
